src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingStructureBuilder.java

```diff
@@ -0,0 +1,332 @@
+            final ItemStack itemStack = new ItemStack(resource.getItem(), resource.getAmount(), resource.getDamageValue());
+            itemStack.setTagCompound(resource.getItemStack().getTagCompound());
+            itemStack.writeToNBT(neededRes);
+
+            neededResTagList.appendTag(neededRes);
+        }
+        compound.setTag(TAG_RESOURCE_LIST, neededResTagList);
+    }
+
+    /**
+     * Method to serialize data to send it to the view.
+     *
+     * @param buf the used ByteBuffer.
+     */
+    @Override
+    public void serializeToView(@NotNull final ByteBuf buf)
+    {
+        super.serializeToView(buf);
+
+        updateAvailableResources();
+        buf.writeInt(neededResources.size());
+        for (@NotNull final BuildingBuilderResource resource : neededResources.values())
+        {
+            ByteBufUtils.writeItemStack(buf, resource.getItemStack());
+            buf.writeInt(resource.getAvailable());
+            buf.writeInt(resource.getAmount());
+        }
+
+        final CitizenData data = this.getMainWorker();
+        if(data != null && data.getJob() instanceof AbstractJobStructure)
+        {
+            final AbstractJobStructure structureBuilderJob = (AbstractJobStructure) data.getJob();
+            final WorkOrderBuildDecoration workOrderBuildDecoration = structureBuilderJob.getWorkOrder();
+            if(workOrderBuildDecoration != null)
+            {
+                final BlockPos pos = workOrderBuildDecoration.getBuildingLocation();
+                final String name =
+                        workOrderBuildDecoration instanceof WorkOrderBuild ? ((WorkOrderBuild) workOrderBuildDecoration).getUpgradeName() : workOrderBuildDecoration.getName();
+                ByteBufUtils.writeUTF8String(buf, name);
+
+                final String desc;
+                if(pos.equals(getLocation()))
+                {
+                    desc = "here";
+                }
+                else
+                {
+                    final BlockPos relativePos = getLocation().subtract(pos);
+                    final EnumFacing facingX = EnumFacing.getFacingFromVector(relativePos.getX(), 0, 0);
+                    final EnumFacing facingZ = EnumFacing.getFacingFromVector(0, 0, relativePos.getZ());
+                    desc = relativePos.getX() + " " + facingX + " " + relativePos.getZ() + " " + facingZ;
+                }
+
+                ByteBufUtils.writeUTF8String(buf, desc);
+            }
+            else
+            {
+                ByteBufUtils.writeUTF8String(buf, "-");
+                ByteBufUtils.writeUTF8String(buf, "");
+            }
+        }
+        else
+        {
+            ByteBufUtils.writeUTF8String(buf, "-");
+            ByteBufUtils.writeUTF8String(buf, "");
+        }
+    }
+
+    /**
+     * Update the available resources.
+     * <p>
+     * which are needed for the build and in the structureBuilder's chest or inventory
+     */
+    private void updateAvailableResources()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 22 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)